Crash on starred expression (*expr within a list expression)

```
[1, *[2, 3]]
```

```
Traceback (most recent call last):
  File "/home/dominik/.virtualenvs/mypy/bin/mypy", line 6, in <module>
    main(__file__)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/main.py", line 54, in main
    res = type_check_only(sources, bin_dir, options)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/main.py", line 102, in type_check_only
    python_path=options.python_path)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 211, in build
    dispatch(sources, manager)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1336, in dispatch
    process_graph(graph, manager)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1467, in process_graph
    process_stale_scc(graph, scc)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1497, in process_stale_scc
    graph[id].type_check()
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1316, in type_check
    manager.type_checker.visit_file(self.tree, self.xpath)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 426, in visit_file
    self.accept(d)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 476, in accept
    typ = node.accept(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/nodes.py", line 729, in accept
    return visitor.visit_expression_stmt(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 1582, in visit_expression_stmt
    self.accept(s.expr)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 476, in accept
    typ = node.accept(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/nodes.py", line 1434, in accept
    return visitor.visit_list_expr(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 2122, in visit_list_expr
    return self.expr_checker.visit_list_expr(e)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 1251, in visit_list_expr
    e)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 1270, in check_list_or_set_expr
    [nodes.ARG_POS] * len(items), context)[0]
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 242, in check_call
    messages=arg_messages)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 661, in check_argument_types
    actual + 1, i + 1, callee, context, messages)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 686, in check_arg
    elif not is_subtype(caller_type, callee_type):
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/subtypes.py", line 49, in is_subtype
    return left.accept(SubtypeVisitor(right, type_parameter_checker))
AttributeError: 'NoneType' object has no attribute 'accept'
```

```
$ python --version
Python 3.5.0+
$ mypy --version
mypy 0.4.2
```

Just installed mypy from PyPI and it produces the same traceback.

```
(mypy) dominik@ubuntu:/tmp$ python -V
Python 3.5.2
(mypy) dominik@ubuntu:/tmp$ apt-cache policy python3.5
python3.5:
  Installed: 3.5.2-2~16.01
  Candidate: 3.5.2-2~16.01
  Version table:
 *** 3.5.2-2~16.01 500
        500 http://pl.archive.ubuntu.com/ubuntu xenial-updates/main amd64 Packages
        100 /var/lib/dpkg/status
     3.5.1-10 500
        500 http://pl.archive.ubuntu.com/ubuntu xenial/main amd64 Packages
(mypy) $ mypy -V
mypy 0.4.3
(mypy) dominik@ubuntu:/tmp$ echo '[1, *[2, 3]]' > main.py
(mypy) dominik@ubuntu:/tmp$ mypy main.py
Traceback (most recent call last):
    main(__file__)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/main.py", line 40, in main
    res = type_check_only(sources, bin_dir, options)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/main.py", line 81, in type_check_only
    options=options)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 177, in build
    dispatch(sources, manager)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1323, in dispatch
    process_graph(graph, manager)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1461, in process_graph
    process_stale_scc(graph, scc)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1538, in process_stale_scc
    graph[id].type_check()
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1301, in type_check
    manager.type_checker.visit_file(self.tree, self.xpath)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 152, in visit_file
    self.accept(d)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 201, in accept
    typ = node.accept(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/nodes.py", line 737, in accept
    return visitor.visit_expression_stmt(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 1333, in visit_expression_stmt
    self.accept(s.expr)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 201, in accept
    typ = node.accept(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/nodes.py", line 1445, in accept
    return visitor.visit_list_expr(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 1859, in visit_list_expr
    return self.expr_checker.visit_list_expr(e)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 1283, in visit_list_expr
    e)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 1303, in check_list_or_set_expr
    [nodes.ARG_POS] * len(items), context)[0]
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 253, in check_call
    messages=arg_messages)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 677, in check_argument_types
    actual + 1, i + 1, callee, context, messages)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 702, in check_arg
    elif not is_subtype(caller_type, callee_type):
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/subtypes.py", line 51, in is_subtype
    return left.accept(SubtypeVisitor(right, type_parameter_checker))
AttributeError: 'NoneType' object has no attribute 'accept'

*** INTERNAL ERROR ***

main.py:1: error: Internal error -- please report a bug at https://github.com/python/mypy/issues

NOTE: you can use "mypy --pdb ..." to drop into the debugger when this happens.
```

`(1, *(2, 3))` crashes with different traceback:

```
(mypy) dominik@ubuntu:/tmp$ echo '(1, *(2, 3))' > main.py
(mypy) dominik@ubuntu:/tmp$ mypy main.py
Traceback (most recent call last):
  File "/home/dominik/.virtualenvs/mypy/bin/mypy", line 6, in <module>
    main(__file__)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/main.py", line 40, in main
    res = type_check_only(sources, bin_dir, options)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/main.py", line 81, in type_check_only
    options=options)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 177, in build
    dispatch(sources, manager)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1323, in dispatch
    process_graph(graph, manager)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1461, in process_graph
    process_stale_scc(graph, scc)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1538, in process_stale_scc
    graph[id].type_check()
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/build.py", line 1301, in type_check
    manager.type_checker.visit_file(self.tree, self.xpath)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 152, in visit_file
    self.accept(d)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 201, in accept
    typ = node.accept(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/nodes.py", line 737, in accept
    return visitor.visit_expression_stmt(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 1333, in visit_expression_stmt
    self.accept(s.expr)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 201, in accept
    typ = node.accept(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/nodes.py", line 1477, in accept
    return visitor.visit_tuple_expr(self)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checker.py", line 1865, in visit_tuple_expr
    return self.expr_checker.visit_tuple_expr(e)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/checkexpr.py", line 1324, in visit_tuple_expr
    fallback_item = join.join_type_list(items)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/join.py", line 358, in join_type_list
    joined = join_types(joined, t)
  File "/home/dominik/.virtualenvs/mypy/lib/python3.5/site-packages/mypy/join.py", line 73, in join_types
    return t.accept(TypeJoinVisitor(s))
AttributeError: 'NoneType' object has no attribute 'accept'

*** INTERNAL ERROR ***

main.py:1: error: Internal error -- please report a bug at https://github.com/python/mypy/issues

NOTE: you can use "mypy --pdb ..." to drop into the debugger when this happens.
```

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -14,7 +14,7 @@ from mypy.nodes import (
     TupleExpr, DictExpr, FuncExpr, SuperExpr, SliceExpr, Context,
     ListComprehension, GeneratorExpr, SetExpr, MypyFile, Decorator,
     ConditionalExpr, ComparisonExpr, TempNode, SetComprehension,
-    DictionaryComprehension, ComplexExpr, EllipsisExpr,
+    DictionaryComprehension, ComplexExpr, EllipsisExpr, StarExpr,
     TypeAliasExpr, BackquoteExpr, ARG_POS, ARG_NAMED, ARG_STAR2
 )
 from mypy.nodes import function_type
@@ -1306,28 +1306,45 @@ class ExpressionChecker:
             name=tag,
             variables=[tvdef])
         return self.check_call(constructor,
-                               items,
-                               [nodes.ARG_POS] * len(items), context)[0]
+                               [(i.expr if isinstance(i, StarExpr) else i)
+                                for i in items],
+                               [(nodes.ARG_STAR if isinstance(i, StarExpr) else nodes.ARG_POS)
+                                for i in items],
+                               context)[0]
 
     def visit_tuple_expr(self, e: TupleExpr) -> Type:
         """Type check a tuple expression."""
         ctx = None  # type: TupleType
-        # Try to determine type context for type inference.
-        if isinstance(self.chk.type_context[-1], TupleType):
-            t = self.chk.type_context[-1]
-            if len(t.items) == len(e.items):
-                ctx = t
+        if not any(isinstance(i, StarExpr) for i in e.items):
+            # Try to determine type context for type inference.
+            if isinstance(self.chk.type_context[-1], TupleType):
+                t = self.chk.type_context[-1]
+                if len(t.items) == len(e.items):
+                    ctx = t
         # Infer item types.
+        staritems = []  # type: List[Type]
         items = []  # type: List[Type]
         for i in range(len(e.items)):
             item = e.items[i]
             tt = None  # type: Type
-            if not ctx:
-                tt = self.accept(item)
+            if isinstance(item, StarExpr):
+                assert ctx is None
+                tt = self.accept(item.expr)
+                self.check_not_void(tt, e)
+                if isinstance(tt, TupleType):
+                    items.extend(tt.items)
+                else:
+                    staritems.append(tt)
             else:
-                tt = self.accept(item, ctx.items[i])
-            self.check_not_void(tt, e)
-            items.append(tt)
+                if not ctx:
+                    tt = self.accept(item)
+                else:
+                    tt = self.accept(item, ctx.items[i])
+                self.check_not_void(tt, e)
+                items.append(tt)
+        if staritems:
+            # XXX Should do better!
+            return self.chk.named_generic_type('builtins.tuple', [AnyType()])
         fallback_item = join.join_type_list(items)
         return TupleType(items, self.chk.named_generic_type('builtins.tuple', [fallback_item]))
 

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1286,15 +1286,17 @@ class ExpressionChecker:
 
     def visit_list_expr(self, e: ListExpr) -> Type:
         """Type check a list expression [...]."""
-        return self.check_list_or_set_expr(e.items, 'builtins.list', '<list>',
-                                           e)
+        return self.check_lst_expr(e.items, 'builtins.list', '<list>', e)
 
     def visit_set_expr(self, e: SetExpr) -> Type:
-        return self.check_list_or_set_expr(e.items, 'builtins.set', '<set>', e)
+        return self.check_lst_expr(e.items, 'builtins.set', '<set>', e)
 
-    def check_list_or_set_expr(self, items: List[Node], fullname: str,
-                               tag: str, context: Context) -> Type:
+    def check_lst_expr(self, items: List[Node], fullname: str,
+                       tag: str, context: Context) -> Type:
         # Translate into type checking a generic function call.
+        # Used for list and set expressions, as well as for tuples
+        # containing star expressions that don't refer to a
+        # Tuple. (Note: "lst" stands for list-set-tuple. :-)
         tvdef = TypeVarDef('T', -1, [], self.chk.object_type())
         tv = TypeVarType(tvdef)
         constructor = CallableType(
@@ -1315,36 +1317,45 @@ class ExpressionChecker:
     def visit_tuple_expr(self, e: TupleExpr) -> Type:
         """Type check a tuple expression."""
         ctx = None  # type: TupleType
-        if not any(isinstance(i, StarExpr) for i in e.items):
-            # Try to determine type context for type inference.
-            if isinstance(self.chk.type_context[-1], TupleType):
-                t = self.chk.type_context[-1]
-                if len(t.items) == len(e.items):
-                    ctx = t
-        # Infer item types.
-        staritems = []  # type: List[Type]
+        # Try to determine type context for type inference.
+        if isinstance(self.chk.type_context[-1], TupleType):
+            t = self.chk.type_context[-1]
+            ctx = t
+        # NOTE: it's possible for the context to have a different
+        # number of items than e.  In that case we use those context
+        # items that match a position in e, and we'll worry about type
+        # mismatches later.
+
+        # Infer item types.  Give up if there's a star expression
+        # that's not a Tuple.
         items = []  # type: List[Type]
+        j = 0  # Index into ctx.items; irrelevant if ctx is None.
         for i in range(len(e.items)):
             item = e.items[i]
             tt = None  # type: Type
             if isinstance(item, StarExpr):
-                assert ctx is None
+                # Special handling for star expressions.
+                # TODO: If there's a context, and item.expr is a
+                # TupleExpr, flatten it, so we can benefit from the
+                # context?  Counterargument: Why would anyone write
+                # (1, *(2, 3)) instead of (1, 2, 3) except in a test?
                 tt = self.accept(item.expr)
                 self.check_not_void(tt, e)
                 if isinstance(tt, TupleType):
                     items.extend(tt.items)
+                    j += len(tt.items)
                 else:
-                    staritems.append(tt)
+                    # A star expression that's not a Tuple.
+                    # Treat the whole thing as a variable-length tuple.
+                    return self.check_lst_expr(e.items, 'builtins.tuple', '<tuple>', e)
             else:
-                if not ctx:
+                if not ctx or j >= len(ctx.items):
                     tt = self.accept(item)
                 else:
-                    tt = self.accept(item, ctx.items[i])
+                    tt = self.accept(item, ctx.items[j])
+                    j += 1
                 self.check_not_void(tt, e)
                 items.append(tt)
-        if staritems:
-            # XXX Should do better!
-            return self.chk.named_generic_type('builtins.tuple', [AnyType()])
         fallback_item = join.join_type_list(items)
         return TupleType(items, self.chk.named_generic_type('builtins.tuple', [fallback_item]))
 

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -1919,14 +1919,20 @@ class SemanticAnalyzer(NodeVisitor):
 
     def visit_tuple_expr(self, expr: TupleExpr) -> None:
         for item in expr.items:
+            if isinstance(item, StarExpr):
+                item.valid = True
             item.accept(self)
 
     def visit_list_expr(self, expr: ListExpr) -> None:
         for item in expr.items:
+            if isinstance(item, StarExpr):
+                item.valid = True
             item.accept(self)
 
     def visit_set_expr(self, expr: SetExpr) -> None:
         for item in expr.items:
+            if isinstance(item, StarExpr):
+                item.valid = True
             item.accept(self)
 
     def visit_dict_expr(self, expr: DictExpr) -> None:
@@ -1936,6 +1942,7 @@ class SemanticAnalyzer(NodeVisitor):
 
     def visit_star_expr(self, expr: StarExpr) -> None:
         if not expr.valid:
+            # XXX TODO Change this error message
             self.fail('Can use starred expression only as assignment target', expr)
         else:
             expr.expr.accept(self)

--- a/test-data/unit/check-inference.test
+++ b/test-data/unit/check-inference.test
@@ -809,6 +809,14 @@ s = s_i()
 s = s_s() # E: Incompatible types in assignment (expression has type Set[str], variable has type Set[int])
 [builtins fixtures/set.py]
 
+[case testSetWithStarExpr]
+# options: fast_parser
+s = {1, 2, *(3, 4)}
+t = {1, 2, *s}
+reveal_type(s)  # E: Revealed type is 'builtins.set[builtins.int*]'
+reveal_type(t)  # E: Revealed type is 'builtins.set[builtins.int*]'
+[builtins fixtures/set.py]
+
 
 -- For statements
 -- --------------

--- a/test-data/unit/check-lists.test
+++ b/test-data/unit/check-lists.test
@@ -69,4 +69,6 @@ a = [1, *[2, 3]]
 reveal_type(a)  # E: Revealed type is 'builtins.list[builtins.int]'
 b = [0, *a]
 reveal_type(b)  # E: Revealed type is 'builtins.list[builtins.int*]'
+c = [*a, 0]
+reveal_type(c)  # E: Revealed type is 'builtins.list[builtins.int*]'
 [builtins fixtures/list.py]

--- a/test-data/unit/check-lists.test
+++ b/test-data/unit/check-lists.test
@@ -63,6 +63,10 @@ class C: pass
 [out]
 main: note: In function "f":
 
-[case testListContainingStarExpr]
-a = [1, *[2]]  # E: Can use starred expression only as assignment target
+[case testListWithStarExpr]
+(x, *a) = [1, 2, 3]
+a = [1, *[2, 3]]
+reveal_type(a)  # E: Revealed type is 'builtins.list[builtins.int]'
+b = [0, *a]
+reveal_type(b)  # E: Revealed type is 'builtins.list[builtins.int*]'
 [builtins fixtures/list.py]

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -768,3 +768,9 @@ a = ()
 from typing import Sized
 a = None  # type: Sized
 a = ()
+
+[case testTupleWithStarExpr]
+# options: fast_parser
+a = (1, 2)
+b = (*a, '', '')
+reveal_type(b)  # E: Revealed type is 'Tuple[builtins.int, builtins.int, builtins.str, builtins.str]'

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -785,6 +785,8 @@ reveal_type(b)  # E: Revealed type is 'builtins.tuple[builtins.int*]'
 a = ['']
 b = (0, *a)
 reveal_type(b)  # E: Revealed type is 'builtins.tuple[builtins.object*]'
+c = (*a, '')
+reveal_type(c)  # E: Revealed type is 'builtins.tuple[builtins.str*]'
 [builtins fixtures/tuple.py]
 
 [case testTupleWithStarExpr4]

--- a/test-data/unit/check-tuples.test
+++ b/test-data/unit/check-tuples.test
@@ -769,8 +769,40 @@ from typing import Sized
 a = None  # type: Sized
 a = ()
 
-[case testTupleWithStarExpr]
+[case testTupleWithStarExpr1]
 # options: fast_parser
 a = (1, 2)
-b = (*a, '', '')
-reveal_type(b)  # E: Revealed type is 'Tuple[builtins.int, builtins.int, builtins.str, builtins.str]'
+b = (*a, '')
+reveal_type(b)  # E: Revealed type is 'Tuple[builtins.int, builtins.int, builtins.str]'
+
+[case testTupleWithStarExpr2]
+a = [1]
+b = (0, *a)
+reveal_type(b)  # E: Revealed type is 'builtins.tuple[builtins.int*]'
+[builtins fixtures/tuple.py]
+
+[case testTupleWithStarExpr3]
+a = ['']
+b = (0, *a)
+reveal_type(b)  # E: Revealed type is 'builtins.tuple[builtins.object*]'
+[builtins fixtures/tuple.py]
+
+[case testTupleWithStarExpr4]
+a = (1, 1, 'x', 'x')
+b = (1, 'x')
+a = (0, *b, '')
+[builtins fixtures/tuple.py]
+
+[case testTupleWithUndersizedContext]
+a = ([1], 'x')
+a = ([], 'x', 1)  # E: Incompatible types in assignment (expression has type "Tuple[List[int], str, int]", variable has type "Tuple[List[int], str]")
+[builtins fixtures/tuple.py]
+
+[case testTupleWithOversizedContext]
+a = (1, [1], 'x')
+a = (1, [])  # E: Incompatible types in assignment (expression has type "Tuple[int, List[int]]", variable has type "Tuple[int, List[int], str]")
+[builtins fixtures/tuple.py]
+
+[case testTupleWithoutContext]
+a = (1, [])  # E: Need type annotation for variable
+[builtins fixtures/tuple.py]

--- a/test-data/unit/fixtures/tuple.py
+++ b/test-data/unit/fixtures/tuple.py
@@ -1,6 +1,6 @@
 # Builtins stub used in tuple-related test cases.
 
-from typing import Iterable, TypeVar, Generic, Sequence
+from typing import Iterable, Iterator, TypeVar, Generic, Sequence
 
 Tco = TypeVar('Tco', covariant=True)
 
@@ -11,6 +11,7 @@ class type:
     def __init__(self, *a) -> None: pass
     def __call__(self, *a) -> object: pass
 class tuple(Sequence[Tco], Generic[Tco]):
+    def __iter__(self) -> Iterator[Tco]: pass
     def __getitem__(self, x: int) -> Tco: pass
 class function: pass
 

--- a/test-data/unit/fixtures/tuple.py
+++ b/test-data/unit/fixtures/tuple.py
@@ -21,6 +21,8 @@ class str: pass # For convenience
 
 T = TypeVar('T')
 
+class list(Sequence[T], Generic[T]): pass
+
 def sum(iterable: Iterable[T], start: T = None) -> T: pass
 
 True = bool()

--- a/test-data/unit/lib-stub/typing.py
+++ b/test-data/unit/lib-stub/typing.py
@@ -70,7 +70,7 @@ class AsyncIterator(AsyncIterable[T], Generic[T]):
     @abstractmethod
     def __anext__(self) -> Awaitable[T]: pass
 
-class Sequence(Generic[T]):
+class Sequence(Iterable[T], Generic[T]):
     @abstractmethod
     def __getitem__(self, n: Any) -> T: pass
 

--- a/test-data/unit/semanal-errors.test
+++ b/test-data/unit/semanal-errors.test
@@ -486,12 +486,8 @@ b = 1
 c = 1
 d = 1
 a = *b
-a = b, (c, *d)
-a = [1, *[2]]
 [out]
 main:4: error: Can use starred expression only as assignment target
-main:5: error: Can use starred expression only as assignment target
-main:6: error: Can use starred expression only as assignment target
 
 [case testStarExpressionInExp]
 a = 1

